server/graphql/v2/mutation/PaymentMethodMutations.js

```diff
@@ -1,61 +1,76 @@
-    description: 'Add a new payment method to be used with an Order',
-    args: {
-      paymentMethod: {
-        type: PaymentMethodCreateInput,
-        description: 'A Payment Method to add to an Account',
-      },
-      account: {
-        type: AccountReferenceInput,
-        description: 'Account to add Payment Method to',
-      },
+const addCreditCard = {
+  type: new GraphQLNonNull(PaymentMethod),
+  description: 'Add a new payment method to be used with an Order',
+  args: {
+    paymentMethod: {
+      type: PaymentMethodCreateInput,
+      description: 'A Payment Method to add to an Account',
+      deprecationReason: '2020-08-24: Please use creditCardInfo',
     },
-    async resolve(_, args) {
-      const account = await fetchAccountWithReference(args.account);
-      const collective = await models.Collective.findByPk(account.id);
-      if (!collective) {
-        throw Error('This collective does not exist');
-      }
-
-      const { paymentMethod } = args;
+    creditCardInfo: {
+      type: CreditCardCreateInput,
+      description: 'The credit card info',
+    },
+    name: {
+      type: GraphQLString,
+      description: 'Name associated to this credit card',
+    },
+    isSavedForLater: {
+      type: GraphQLBoolean,
+      description: 'Wether this payment method should be saved for future payments',
+    },
+    account: {
+      type: new GraphQLNonNull(AccountReferenceInput),
+      description: 'Account to add Payment Method to',
+    },
+  },
+  async resolve(_, args, req) {
+    const collective = await fetchAccountWithReference(args.account, { throwIfMissing: true });
+    if (!req.remoteUser.isAdminOfCollective(collective)) {
+      throw new Forbidden(`Must be an admin of ${collective.name}`);
+    }
 
-      const newPaymentMethodData = {
-        ...paymentMethod,
-        service: 'stripe',
-        type: 'creditcard',
-        CreatedByUserId: account.CreatedByUserId,
-        currency: account.currency,
-        saved: true,
-        CollectiveId: account.id,
-      };
+    const newPaymentMethodData = {
+      service: 'stripe',
+      type: 'creditcard',
+      name: args.name,
+      CreatedByUserId: req.remoteUser.id,
+      currency: collective.currency,
+      saved: args.isSavedForLater,
+      CollectiveId: collective.id,
+    };
 
-      let pm = await models.PaymentMethod.create(newPaymentMethodData);
+    if (args.creditCardInfo) {
+      const data = pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']);
+      Object.assign(newPaymentMethodData, { token: args.creditCardInfo.token, data });
+    } else if (args.paymentMethod) {
+      Object.assign(newPaymentMethodData, pick(args.paymentMethod, ['data', 'name', 'token']));
+    } else {
+      throw new ValidationFailed('Either creditCardInfo or paymentMethod must be provided');
+    }
 
-      try {
-        pm = await setupCreditCard(pm, {
-          collective,
-          user: account,
```

Comment:
We were passing a `collective` to the `user` field, but this was only used for the `user.email`. As a consequence, the "customers" created in https://github.com/opencollective/opencollective-api/blob/c54ac505ec3718ea6e0f4de2cabfdd6e9707805e/server/paymentProviders/stripe/creditcard.js#L28 have been setup without an email. @znarf is that an issue?

Reply:
Good to know. Should not cause any issue.